### DISCUSSION TITLE
Renovate upgrades Ruby to a version that it can't handle

### DISCUSSION THREAD

**Discussion by jimbali** [renovatebot/renovate#10194] [Request Help]
**How are you running Renovate?**

- [ ] WhiteSource Renovate hosted app on github.com
- [x] Self hosted

If using the hosted app, please skip to the next section.
Otherwise, if self-hosted, please complete the following:

Please select which platform you are using:

- [ ] Azure DevOps (dev.azure.com)
- [ ] Azure DevOps Server
- [ ] Bitbucket Cloud (bitbucket.org)
- [x] Bitbucket Server
- [ ] Gitea
- [ ] github.com
- [ ] GitHub Enterprise Server
- [ ] gitlab.com
- [ ] GitLab self-hosted

Renovate version: 25.31.10 (from Docker Hub)

**Describe the bug**

Renovate Bot runs against my Ruby 3.0.0 project and raises a PR to upgrade to Ruby 3.0.1. When I merge that PR and run Renovate Bot again (using the Docker image), any PRs it tries to create have an error saying that it cannot find Ruby version 3.0.1 in the source.

I suppose this means that the Docker image doesn't have that Ruby version? Really I don't think Renovate should be raising PRs to update to a version that it doesn't support, but ideally it should always support the latest Ruby (and a number of older versions).

**Relevant debug logs**

<details><summary>Click me to see logs</summary>

```
Fetching gem metadata from https://nexus.tooling.local:8443/repository/ruby-gems/....
Fetching gem metadata from https://nexus.tooling.local:8443/repository/ruby-gems/..
Fetching gem metadata from https://rubygems.org/.........
Fetching gem metadata from https://nexus.tooling.local:8443/repository/ruby-gems/..
Fetching gem metadata from https://nexus.tooling.local:8443/repository/ruby-gems/...
Fetching gem metadata from https://rubygems.org/..
Fetching gem metadata from https://nexus.tooling.local:8443/repository/ruby-gems/..
Fetching gem metadata from https://nexus.tooling.local:8443/repository/ruby-gems/...
Fetching gem metadata from https://rubygems.org/..

Could not find gem 'Ruby (~> 3.0.1.0)' in the local ruby installation.
The source contains the following versions of 'Ruby': 3.0.0.0
```

</details>

**Have you created a minimal reproduction repository?**

Please read the [minimal reproductions documentation](https://github.com/renovatebot/renovate/blob/main/docs/development/minimal-reproductions.md) to learn how to make a good minimal reproduction repository.

- [ ] I have provided a minimal reproduction repository
- [ ] I don't have time for that, but it happens in a public repository I have linked to
- [x] I don't have time for that, and cannot share my private repository
- [ ] The nature of this bug means it's impossible to reproduce publicly

**Additional context**

...

**Comment by viceice** [comment]
The renovate full image will always be delayed with global tools update.

https://github.com/renovatebot/docker-renovate-full/blob/61809a5071435eca7da27d1eb0e76527a2fa1580/Dockerfile#L82

Also renovate didn't know which versions are globally installed, as this can be done by any user.

So we suggest to run the [slim](https://github.com/renovatebot/docker-renovate#docker-renovate) renovate image with `binarySource=docker`, so renovate can run the required tools versions as docker sidecar containers.

The slim image is also much smaller, as it has no global tools beside node and renovate installed.

**Reply by jimbali** [reply]
ok, but that means that renovate will not work properly if a PR is merged to update the version. It's been like this for several days so I'm guessing that it takes a while for the tools to be updated. So what is the best course of action if a PR is raised to update Ruby? Options are:

* Just wait and don't merge the PR. That requires checking Renovate each day to see if it has the version of Ruby yet. During that time, the project could be vulnerable.
* Renovate could include a configuration option to prevent updates to Ruby versions that it doesn't support. This would make Renovate work correctly but would also leave the project potentially vulnerable.
* Renovate could have a warning in the PR that if it is merged, Renovate may not work correctly. This means that the project would be kept up-to-date but Renovate would be broken periodically. Not ideal but at least it would be clearer.

Personally I'm not sure what the best solution is. How long does it normally take for a new Ruby version to be available in the image?

**Reply by jimbali** [reply]
oh I see, I just saw your updated comment about the slim image and sidecar containers with the correct tooling. I'll see if I can get it to work that way. Thanks

**Reply by viceice** [reply]
Yeah, sorry, github has add my comment before i could finish. We will deprecate the full image in near future and provide a simple dockerfile for users to build theire own full image, if they can't use the docker sidecar approach.

You can also use Docker-in-Docker. This is the way our renovate template works on gitlab, https://gitlab.com/renovate-bot/renovate-runner

**Reply by viceice** [reply]
btw: here is the ruby sidecar image already on v3.0.1 https://hub.docker.com/r/renovate/ruby/tags?page=1&ordering=last_updated

**Reply by viceice** [reply]
I also found a bug in the full dockerfile, thats why ruby isn't updated yet, fixing it now

https://github.com/renovatebot/docker-renovate-full/commit/63e50cfa189203fcbf5c512cceffdd99d3f29573

**Reply by jimbali** [reply]
nice one, thanks!

**Reply by jimbali** [reply]
Hi @viceice. Do you have any examples of how to run Renovate with the sidecar pattern please? I tried this in OpenShift with the renovate/renovate:slim image in one container and the renovate/ruby image in another container within the pod, but unfortunately I hit a wall in that our OpenShift environment has no wide internet connection so it's not really useful in this case. How can the same pattern be achieved with either standalone Docker (docker-compose? links between containers?) or Docker-in-Docker (which I am not really familiar with)?

I suppose our other option is to build our own container based on the slim image (which we already do in order to bake in our CA certs) - would we be able to build one using something like rbenv or rvm so that it could handle multiple Ruby versions?

**Reply by viceice** [reply]
i'm not sure if rvm or rbenv are helping, you can of cause add global tools to the slim image, but you need to add `ENV RENOVATE_BINARY_SOURCE=global`

checkout [Dockerfile](https://github.com/renovatebot/docker-renovate-full/blob/main/Dockerfile) to see how you can install custom tools for renovate.

Here is a sample for a gitlab dind pipeline: https://gitlab.com/renovate-bot/renovate-runner/-/blob/main/templates/renovate-dind.gitlab-ci.yml


We are also preparing a pr for our helm-chart for renovate with dind support: https://github.com/renovatebot/helm-charts/pull/131

**Reply by jimbali** [reply]
@viceice I think the problem I'm actually trying to solve is that our Gemfiles have Ruby locked to a specific version, equal to that specified in the `.ruby-version` file.
```
ruby File.read(File.join(__dir__, '.ruby-version').strip)
```
So when Renovate updates the ruby version, it then becomes unable to run Bundler because Renovate has a different ruby version than the one specified in the Gemfile.

So does Renovate support multiple versions of ruby so that you can run it against different projects with different ruby versions? Or do we need to run different instances of Renovate for each version of Ruby that we support?

**Reply by viceice** [reply]
as i said, a single renovate image can only have one global ruby version. so your best solution is to use the docker dind feature.

you can use `dockerImagePrefix` tto point renovate to an internal docker registry, there you need to create a proxy to docker hub or push the required renovate images manually or from some other ci job.

I'm using the official `registry` image as proxy for renovate to not get blocked by official hub limits. 

https://docs.renovatebot.com/self-hosted-configuration/#dockerimageprefix
https://docs.docker.com/registry/deploying/

btw: bundler manager uses `.ruby-version` file too to find the required ruby version:
https://github.com/renovatebot/renovate/blob/3be44469bb98600a8f839887f9cb3c5c6bf975f3/lib/manager/bundler/artifacts.ts#L29-L55

https://github.com/renovatebot/renovate/blob/b8e36daa87e4275caa5b66bffa6fe3f4d5d8a77e/lib/manager/bundler/extract.ts#L32-L40
